File: sdk/swift/Sources/Senda/MeshClient.swift
import Foundation

public struct InviteToken: Sendable {
    public let value: String

    public init(_ value: String) {
        self.value = value
    }
}

public struct Model: Sendable {
    public let id: String
    public let name: String
}

public struct MeshStatus: Sendable {
    public let connected: Bool
    public let peerCount: Int
}

public struct RequestId: Sendable {
    public let value: String
}

public enum MeshEvent: Sendable {
    case connecting
    case joined(nodeId: String)
    case modelsUpdated(models: [Model])
    case tokenDelta(requestId: String, delta: String)
    case completed(requestId: String)
    case failed(requestId: String, error: String)
    case disconnected(reason: String)
}

public struct ChatMessage: Sendable {
    public let role: String
    public let content: String

    public init(role: String, content: String) {
        self.role = role
        self.content = content
    }
}

public struct ChatRequest: Sendable {
    public let model: String
    public let messages: [ChatMessage]

    public init(model: String, messages: [ChatMessage]) {
        self.model = model
        self.messages = messages
    }
}

public struct ResponsesRequest: Sendable {
    public let model: String
    public let input: String

    public init(model: String, input: String) {
        self.model = model
        self.input = input
    }
}

#if canImport(SendaFFI)
public final class MeshClient: @unchecked Sendable {
    private let inviteToken: InviteToken
    private let ownerKeypairBytesHex: String
    private let stateLock = NSLock()
    private var handle: MeshClientHandle?

    public init(inviteToken: InviteToken, ownerKeypairBytesHex: String = "") {
        self.inviteToken = inviteToken
        self.ownerKeypairBytesHex = ownerKeypairBytesHex
    }

    public func join() async throws {
        let handle = try ensureHandle()
        try await runBlocking {
            try handle.join()
        }
    }

    public func listModels() async throws -> [Model] {
        let handle = try ensureHandle()
        let models = try await runBlocking {
            try handle.listModels()
        }
        return models.map(Self.mapModel)
    }

    public func chat(_ request: ChatRequest) -> AsyncThrowingStream<MeshEvent, Error> {
        let requestDto = Self.mapChatRequest(request)
        return AsyncThrowingStream { continuation in
            do {
                let handle = try self.ensureHandle()
                let bridge = EventStreamBridge(continuation: continuation) { [weak self] requestId in
                    self?.cancel(RequestId(value: requestId))
                }
                let requestId = handle.chat(request: requestDto, listener: bridge)
                bridge.activate(requestId: requestId)
            } catch {
                continuation.finish(throwing: error)
            }
        }
    }

    public func responses(_ request: ResponsesRequest) -> AsyncThrowingStream<MeshEvent, Error> {
        let requestDto = Self.mapResponsesRequest(request)
        return AsyncThrowingStream { continuation in
            do {
                let handle = try self.ensureHandle()
                let bridge = EventStreamBridge(continuation: continuation) { [weak self] requestId in
                    self?.cancel(RequestId(value: requestId))
                }
                let requestId = handle.responses(request: requestDto, listener: bridge)
                bridge.activate(requestId: requestId)
            } catch {
                continuation.finish(throwing: error)
            }
        }
    }

    public func cancel(_ requestId: RequestId) {
        guard let handle = currentHandle() else {
            return
        }
        handle.cancel(requestId: requestId.value)
    }

    public func status() async -> MeshStatus {
        guard let handle = currentHandle() else {
            return MeshStatus(connected: false, peerCount: 0)
        }
        let status = await runBlocking {
            handle.status()
        }
        return MeshStatus(
            connected: status.connected,
            peerCount: Int(clamping: status.peerCount)
        )
    }

    public func disconnect() async {
        guard let handle = currentHandle() else {
            return
        }
        await runBlocking {
            handle.disconnect()
        }
    }

    public func reconnect() async throws {
        let handle = try ensureHandle()
        try await runBlocking {
            try handle.reconnect()
        }
    }

    private func ensureHandle() throws -> MeshClientHandle {
        stateLock.lock()
        defer { stateLock.unlock() }

        if let handle {
            return handle
        }

        let created = try createClient(
            ownerKeypairBytesHex: ownerKeypairBytesHex,
            inviteToken: inviteToken.value
        )
        handle = created
        return created
    }

    private func currentHandle() -> MeshClientHandle? {
        stateLock.lock()
        defer { stateLock.unlock() }
        return handle
    }

    private static func mapModel(_ dto: ModelDto) -> Model {
        Model(id: dto.id, name: dto.name)
    }

    static func mapEvent(_ dto: EventDto) -> MeshEvent {
        switch dto {
        case .connecting:
            return .connecting
        case .joined(let nodeId):
            return .joined(nodeId: nodeId)
        case .modelsUpdated(let models):
            return .modelsUpdated(models: models.map(mapModel))
        case .tokenDelta(let requestId, let delta):
            return .tokenDelta(requestId: requestId, delta: delta)
        case .completed(let requestId):
            return .completed(requestId: requestId)
        case .failed(let requestId, let error):
            return .failed(requestId: requestId, error: error)
        case .disconnected(let reason):
            return .disconnected(reason: reason)
        }
    }

    private static func mapChatRequest(_ request: ChatRequest) -> ChatRequestDto {
        ChatRequestDto(
            model: request.model,
            messages: request.messages.map {
                ChatMessageDto(role: $0.role, content: $0.content)
            }
        )
    }

    private static func mapResponsesRequest(_ request: ResponsesRequest) -> ResponsesRequestDto {
        ResponsesRequestDto(model: request.model, input: request.input)
    }
}
#else
public final class MeshClient: @unchecked Sendable {
    private let inviteToken: InviteToken
    private var isConnected: Bool = false

    public init(inviteToken: InviteToken, ownerKeypairBytesHex _: String = "") {
        self.inviteToken = inviteToken
    }

    public func join() async throws {
        isConnected = true
    }

    public func listModels() async throws -> [Model] {
        []
    }

    public func chat(_ request: ChatRequest) -> AsyncThrowingStream<MeshEvent, Error> {
        let requestId = UUID().uuidString
        return AsyncThrowingStream { continuation in
            continuation.onTermination = { [weak self] _ in
                self?.cancel(RequestId(value: requestId))
            }
            continuation.yield(.completed(requestId: requestId))
            continuation.finish()
        }
    }

    public func responses(_ request: ResponsesRequest) -> AsyncThrowingStream<MeshEvent, Error> {
        let requestId = UUID().uuidString
        return AsyncThrowingStream { continuation in
            continuation.onTermination = { [weak self] _ in
                self?.cancel(RequestId(value: requestId))
            }
            continuation.yield(.completed(requestId: requestId))
            continuation.finish()
        }
    }

    public func cancel(_ requestId: RequestId) {}

    public func status() async -> MeshStatus {
        MeshStatus(connected: isConnected, peerCount: 0)
    }

    public func disconnect() async {
        isConnected = false
    }

    public func reconnect() async throws {
        await disconnect()
        try await join()
    }
}
#endif

private func runBlocking<T>(_ work: @escaping () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                continuation.resume(returning: try work())
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
}

private func runBlocking<T>(_ work: @escaping () -> T) async -> T {
    await withCheckedContinuation { continuation in
        DispatchQueue.global(qos: .userInitiated).async {
            continuation.resume(returning: work())
        }
    }
}

File: sdk/kotlin/src/main/kotlin/com/closedmesh/MeshClient.kt
package com.senda

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.withContext
import uniffi.mesh_ffi.ChatMessageDto
import uniffi.mesh_ffi.ChatRequestDto
import uniffi.mesh_ffi.EventDto
import uniffi.mesh_ffi.EventListener as FfiEventListener
import uniffi.mesh_ffi.MeshClientHandleInterface
import uniffi.mesh_ffi.ModelDto
import uniffi.mesh_ffi.ResponsesRequestDto
import uniffi.mesh_ffi.StatusDto

data class Model(val id: String, val name: String)

data class ChatMessage(val role: String, val content: String)

data class ChatRequest(val model: String, val messages: List<ChatMessage>)

data class ResponsesRequest(val model: String, val input: String)

data class Status(val connected: Boolean, val peerCount: ULong)

@JvmInline
value class RequestId(val value: String)

sealed class Event {
    object Connecting : Event()
    data class Joined(val nodeId: String) : Event()
    data class ModelsUpdated(val models: List<Model>) : Event()
    data class TokenDelta(val requestId: RequestId, val delta: String) : Event()
    data class Completed(val requestId: RequestId) : Event()
    data class Failed(val requestId: RequestId, val error: String) : Event()
    data class Disconnected(val reason: String) : Event()
}

fun interface EventListener {
    fun onEvent(event: Event)
}

private fun ModelDto.toModel() = Model(id = id, name = name)

private fun StatusDto.toStatus() = Status(connected = connected, peerCount = peerCount)

private fun ChatMessage.toDto() = ChatMessageDto(role = role, content = content)

private fun ChatRequest.toDto() =
    ChatRequestDto(model = model, messages = messages.map { it.toDto() })

private fun ResponsesRequest.toDto() = ResponsesRequestDto(model = model, input = input)

private fun EventDto.toEvent(): Event =
    when (this) {
        is EventDto.Connecting -> Event.Connecting
        is EventDto.Joined -> Event.Joined(nodeId = nodeId)
        is EventDto.ModelsUpdated -> Event.ModelsUpdated(models = models.map { it.toModel() })
        is EventDto.TokenDelta -> Event.TokenDelta(requestId = RequestId(requestId), delta = delta)
        is EventDto.Completed -> Event.Completed(requestId = RequestId(requestId))
        is EventDto.Failed -> Event.Failed(requestId = RequestId(requestId), error = error)
        is EventDto.Disconnected -> Event.Disconnected(reason = reason)
    }

class MeshClient(private val handle: MeshClientHandleInterface) {

    suspend fun join(): Unit = withContext(Dispatchers.IO) { handle.join() }

    suspend fun listModels(): List<Model> =
        withContext(Dispatchers.IO) { handle.listModels().map { it.toModel() } }

    fun chat(request: ChatRequest, listener: EventListener): RequestId {
        val bridge = object : FfiEventListener {
            override fun onEvent(event: EventDto) = listener.onEvent(event.toEvent())
        }
        return RequestId(handle.chat(request.toDto(), bridge))
    }

    fun responses(request: ResponsesRequest, listener: EventListener): RequestId {
        val bridge = object : FfiEventListener {
            override fun onEvent(event: EventDto) = listener.onEvent(event.toEvent())
        }
        return RequestId(handle.responses(request.toDto(), bridge))
    }

    fun cancel(requestId: RequestId) = handle.cancel(requestId.value)

    suspend fun status(): Status = withContext(Dispatchers.IO) { handle.status().toStatus() }

    suspend fun disconnect(): Unit = withContext(Dispatchers.IO) { handle.disconnect() }

    suspend fun reconnect(): Unit = withContext(Dispatchers.IO) { handle.reconnect() }

    fun chatFlow(request: ChatRequest): Flow<Event> = callbackFlow {
        var requestId: RequestId? = null
        var terminalRequestId: RequestId? = null
        requestId = chat(request) { event ->
            trySend(event)
            val currentRequestId = requestId
            if (currentRequestId != null && event.isTerminalFor(currentRequestId)) {
                terminalRequestId = currentRequestId
                close()
            } else if (currentRequestId == null) {
                terminalRequestId = event.terminalRequestId()
            }
        }
        if (requestId == terminalRequestId) {
            close()
        }
        awaitClose {
            val currentRequestId = requestId ?: return@awaitClose
            if (terminalRequestId != currentRequestId) {
                cancel(currentRequestId)
            }
        }
    }

    fun responsesFlow(request: ResponsesRequest): Flow<Event> = callbackFlow {
        var requestId: RequestId? = null
        var terminalRequestId: RequestId? = null
        requestId = responses(request) { event ->
            trySend(event)
            val currentRequestId = requestId
            if (currentRequestId != null && event.isTerminalFor(currentRequestId)) {
                terminalRequestId = currentRequestId
                close()
            } else if (currentRequestId == null) {
                terminalRequestId = event.terminalRequestId()
            }
        }
        if (requestId == terminalRequestId) {
            close()
        }
        awaitClose {
            val currentRequestId = requestId ?: return@awaitClose
            if (terminalRequestId != currentRequestId) {
                cancel(currentRequestId)
            }
        }
    }
}

private fun Event.isTerminalFor(requestId: RequestId): Boolean =
    when (this) {
        is Event.Completed -> this.requestId == requestId
        is Event.Failed -> this.requestId == requestId
        else -> false
    }

private fun Event.terminalRequestId(): RequestId? =
    when (this) {
        is Event.Completed -> requestId
        is Event.Failed -> requestId
        else -> null
    }
